services/cd-service/pkg/service/batch.go

```diff
@@ -0,0 +1,139 @@
+		}
+		if !valid.ApplicationName(action.CreateEnvironmentApplicationLock.Application) {
+			return nil, status.Error(codes.InvalidArgument, "invalid application")
+		}
+		if !valid.LockId(action.CreateEnvironmentApplicationLock.LockId) {
+			return nil, status.Error(codes.InvalidArgument, "invalid lock id")
+		}
+		return &repository.CreateEnvironmentApplicationLock{
+			Environment: action.CreateEnvironmentApplicationLock.Environment,
+			Application: action.CreateEnvironmentApplicationLock.Application,
+			LockId:      action.CreateEnvironmentApplicationLock.LockId,
+			Message:     action.CreateEnvironmentApplicationLock.Message,
+		}, nil
+	case *api.BatchAction_DeleteEnvironmentApplicationLock:
+		if !valid.EnvironmentName(action.DeleteEnvironmentApplicationLock.Environment) {
+			return nil, status.Error(codes.InvalidArgument, "invalid environment")
+		}
+		if !valid.ApplicationName(action.DeleteEnvironmentApplicationLock.Application) {
+			return nil, status.Error(codes.InvalidArgument, "invalid application")
+		}
+		if !valid.LockId(action.DeleteEnvironmentApplicationLock.LockId) {
+			return nil, status.Error(codes.InvalidArgument, "invalid lock id")
+		}
+		return &repository.DeleteEnvironmentApplicationLock{
+			Environment: action.DeleteEnvironmentApplicationLock.Environment,
+			Application: action.DeleteEnvironmentApplicationLock.Application,
+			LockId:      action.DeleteEnvironmentApplicationLock.LockId,
+		}, nil
+	case *api.BatchAction_Deploy:
+		if !valid.EnvironmentName(action.Deploy.Environment) {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid environment %q", action.Deploy.Environment))
+		}
+		if !valid.ApplicationName(action.Deploy.Application) {
+			return nil, status.Error(codes.InvalidArgument, fmt.Sprintf("invalid application %q", action.Deploy.Application))
+		}
+		b := action.Deploy.LockBehavior
+		if action.Deploy.IgnoreAllLocks {
+			// the UI currently sets this to true,
+			// in that case, we still want to ignore locks (for emergency deployments)
+			b = api.LockBehavior_Ignore
+		}
+		return &repository.DeployApplicationVersion{
+			Environment:   action.Deploy.Environment,
+			Application:   action.Deploy.Application,
+			Version:       action.Deploy.Version,
+			LockBehaviour: b,
+		}, nil
+	}
+	return nil, status.Error(codes.InvalidArgument, "invalid action")
+}
+
+func (d *BatchServer) ProcessBatch(
+	ctx context.Context,
+	in *api.BatchRequest,
+) (*emptypb.Empty, error) {
+	if len(in.GetActions())>100 {
+		return nil, status.Error(codes.InvalidArgument, "too many actions. limit is 100")
+	}
+	transformers := make ([]repository.Transformer, 0, 100)
+	for _, batchAction := range in.GetActions() {
+		transformer, err := d.processAction(batchAction)
+		if err != nil {
+			// Validation error
+			return nil, err
+		}
+		transformers = append(transformers, transformer)
+	}
+	err := d.Repository.Apply(ctx, transformers...)
+	if err != nil {
+		// TODO TE: error handling
```

Comment:
don't leave "TODOs" in.
I think the best you can do here is wrap the error to add a little bit of context, e.g.
`internalError(ctx, fmt.Errorf("could not apply transformer: %w", err))`